test/src/main/java/org/apache/accumulo/test/ScanConsistencyIT.java

```diff
@@ -0,0 +1,492 @@
+      }
+
+      for (Map.Entry<Key,Value> entry : s) {
+        stats.scanned++;
+        Key key = entry.getKey();
+        key.setTimestamp(0);
+        if (expected.remove(key)) {
+          stats.verified++;
+        }
+      }
+
+      assertTrue(expected.isEmpty());
+    }
+    return stats;
+  }
+
+  private static class ScanTask implements Callable<ScanStats> {
+
+    private final TestContext tctx;
+
+    private ScanTask(TestContext testContext) {
+      this.tctx = testContext;
+    }
+
+    @SuppressFBWarnings(value = {"PREDICTABLE_RANDOM", "DMI_RANDOM_USED_ONLY_ONCE"},
+        justification = "predictable random is ok for testing")
+    @Override
+    public ScanStats call() throws Exception {
+      ScanStats allStats = new ScanStats();
+
+      Random random = new Random();
+
+      while (tctx.keepRunning.get()) {
+
+        Range range;
+        if (random.nextInt(10) == 0) {
+          // 1 in 10 chance of doing a full table scan
+          range = new Range();
+        } else {
+          long start = nextLongAbs(random);
+          long end = nextLongAbs(random);
+
+          while (end <= start) {
+            end = nextLongAbs(random);
+          }
+
+          range = new Range(String.format("%016x", start), String.format("%016x", end));
+        }
+
+        var stats = scanData(tctx, range, random.nextBoolean());
+        allStats.add(stats);
+      }
+
+      return allStats;
+    }
+  }
+
+  private static class WriteStats {
+    long written;
+    long deleted;
+  }
+
+  private static class WriteTask implements Callable<WriteStats> {
+
+    private final TestContext tctx;
+
+    private WriteTask(TestContext testContext) {
+      this.tctx = testContext;
+    }
+
+    @SuppressFBWarnings(value = {"PREDICTABLE_RANDOM", "DMI_RANDOM_USED_ONLY_ONCE"},
+        justification = "predictable random is ok for testing")
+    @Override
+    public WriteStats call() throws Exception {
+      WriteStats stats = new WriteStats();
+
+      Random random = new Random();
+
+      while (tctx.keepRunning.get()) {
+
+        // Each row has on average of 50 cols, so 20K rows would be around 100K entries.
+        if (tctx.dataTracker.estimatedRows() > 100_000 / 50) {
+          // There is a good bit of data, so delete some.
+          try (var writer = tctx.client.createBatchWriter(tctx.table)) {
+            for (Mutation m : tctx.dataTracker.getDeletes()) {
+              writer.addMutation(m);
+              stats.deleted += m.getUpdates().size();
+            }
+          }
+
+        } else {
+
+          List<Mutation> dataAdded = new ArrayList<>();
+
+          int rowsToGenerate = random.nextInt(1000);
+
+          try (var writer = tctx.client.createBatchWriter(tctx.table)) {
+            for (int i = 0; i < rowsToGenerate; i++) {
+              Mutation m = generateMutation(random);
+              writer.addMutation(m);
+              stats.written += m.getUpdates().size();
+              dataAdded.add(m);
+            }
```

Comment:
Is it possible that the BatchWriter flushes in this loop based on default configs? If so, a scan that starts during this loop will see more data than expected.

Reply:
Yeah that is ok, the scans do not care if they see extra data. They just require that they see all data that was successfully written before the scan started.

There is a very last scan done after all the writers have stopped, that scan does not tolerate extra data and should throw an exception if it sees it.